Training/virtual_functions.cpp

```diff
@@ -6,89 +6,140 @@ using namespace std;
-    }
-    void putdata() 
-    {
-        cout << m_name << " " << m_age << " " << m_publications << " " << m_cur_id << endl;
-    }
+	Professor ();
+	static Person * make ();
+	void getdata () override; /// Es-Note: Should explicitly use override keyword.
+	void putdata () override;
 };
 
-int Professor::s_next_id = 1;
 
 class Student : public Person {
 private:
-    static int s_next_id;
-    int m_marks[6] = { -1, -1, -1, -1, -1, -1 };
-    int m_cur_id = -1;
+	static int s_next_id;
+	int m_marks[6] = { -1, -1, -1, -1, -1, -1 };
+	int m_cur_id   = -1;
 public:
-    Student()
-    {
-        m_cur_id = s_next_id++;
-    }
-    void getdata()
-    {
-        cin >> m_name >> m_age;
-        for (int i = 0; i < 6; ++i) {
-            cin >> m_marks[i];
-        }
-    }
-    void putdata()
-    {
-        cout << m_name << " " << m_age;
-        int marks_sum = 0;
-        for (int i = 0; i < 6; ++i) {
-            marks_sum += m_marks[i];
-        }
-        cout << " " << marks_sum << " " << m_cur_id << endl;
-    }
+	Student ();
+	static Person * make ();
+	void getdata () override;
+	void putdata () override;
 };
 
-int Student::s_next_id = 1;
 
-int hackerrank::virtual_functions()
+int hackerrank::virtual_functions ()
 {
-    // START HACKERRANK BOILERPLATE
-    int n, val;
-    cin >> n; //The number of objects that is going to be created.
-    Person *per[1000];
+	// START HACKERRANK BOILERPLATE
+	int n, val;
+	cin >> n; //The number of objects that is going to be created.
+	Person * per[1000];
+	memset(per, 0, sizeof(Person*)); ///Es-Note: Initialize `per` array to nullptrs for safety.
+
+	for (int i = 0; i < n; i++) {
+
+		cin >> val;
+		if (val == 1) {
+			// If val is 1 current object is of type Professor
+			per[i] = new Professor(); ///Es-Note: Use function-call style syntax for constructors.
+
+		}
+		else { ///Es-Note: match brace style w/the if statement-- either both or neither have it.
+			per[i] = new Student(); // Else the current object is of type Student
+		}
+		per[i]->getdata(); // Get the data from the user.
+
+	}
+
+	for (int i = 0; i < n; i++)
+		per[i]->putdata(); // Print the required output for each object.
+
+	return 0;
+	// END HACKERRANK BOILERPLATE
+}
+
+
+///////////////////
+// Professor Impl
 
-    for (int i = 0;i < n;i++) {
+int Professor::s_next_id = 1;
 
-        cin >> val;
-        if (val == 1) {
-            // If val is 1 current object is of type Professor
-            per[i] = new Professor;
+inline Professor::Professor () {
+	m_cur_id = s_next_id++;
+}
+inline void Professor::getdata () {
+	cin >> m_name >> m_age >> m_publications;
+}
+inline void Professor::putdata () {
+	cout << m_name << " " << m_age << " " << m_publications << " " << m_cur_id << endl;
+}
 
-        }
-        else per[i] = new Student; // Else the current object is of type Student
+inline Person * Professor::make () {
+	auto ret = new Professor();
+	ret->m_cur_id = s_next_id;
```

Comment:
`m_cur_id` should be inaccessible here, right?

Reply:
Uhh-- should be fine. The make function is a static member, which means it should have private access like any other member function.

Reply:
`make` is a static member of `Professor`, should be fine.